compiler/rustc_middle/src/ty/consts/valtree.rs

```diff
@@ -3,74 +3,45 @@ use std::ops::Deref;
 
 use rustc_data_structures::intern::Interned;
 use rustc_hir::def::Namespace;
-use rustc_macros::{HashStable, Lift, TyDecodable, TyEncodable, TypeFoldable, TypeVisitable};
+use rustc_macros::{
+    HashStable, Lift, TyDecodable, TyEncodable, TypeFoldable, TypeVisitable, extension,
+};
 
 use super::ScalarInt;
 use crate::mir::interpret::{ErrorHandled, Scalar};
 use crate::ty::print::{FmtPrinter, PrettyPrinter};
 use crate::ty::{self, Ty, TyCtxt};
 
-/// This datastructure is used to represent the value of constants used in the type system.
-///
-/// We explicitly choose a different datastructure from the way values are processed within
-/// CTFE, as in the type system equal values (according to their `PartialEq`) must also have
-/// equal representation (`==` on the rustc data structure, e.g. `ValTree`) and vice versa.
-/// Since CTFE uses `AllocId` to represent pointers, it often happens that two different
-/// `AllocId`s point to equal values. So we may end up with different representations for
-/// two constants whose value is `&42`. Furthermore any kind of struct that has padding will
-/// have arbitrary values within that padding, even if the values of the struct are the same.
-///
-/// `ValTree` does not have this problem with representation, as it only contains integers or
-/// lists of (nested) `ValTree`.
-#[derive(Clone, Debug, Hash, Eq, PartialEq)]
-#[derive(HashStable, TyEncodable, TyDecodable)]
-pub enum ValTreeKind<'tcx> {
-    /// integers, `bool`, `char` are represented as scalars.
-    /// See the `ScalarInt` documentation for how `ScalarInt` guarantees that equal values
-    /// of these types have the same representation.
-    Leaf(ScalarInt),
-
-    //SliceOrStr(ValSlice<'tcx>),
-    // don't use SliceOrStr for now
-    /// The fields of any kind of aggregate. Structs, tuples and arrays are represented by
-    /// listing their fields' values in order.
-    ///
-    /// Enums are represented by storing their variant index as a u32 field, followed by all
-    /// the fields of the variant.
-    ///
-    /// ZST types are represented as an empty slice.
-    Branch(Box<[ValTree<'tcx>]>),
-}
-
-impl<'tcx> ValTreeKind<'tcx> {
+#[extension(pub trait ValTreeKindExt<'tcx>)]
+impl<'tcx> ty::ValTreeKind<TyCtxt<'tcx>> {
     #[inline]
-    pub fn unwrap_leaf(&self) -> ScalarInt {
+    fn unwrap_leaf(&self) -> ScalarInt {
```

Comment:
pls move as many as possible to `rustc_type_ir` instead of adding an extension trait